File: android_app/lib/core/utils/datetime_helper.dart

class DateTimeHelper {
  DateTimeHelper._();

  static const int vietnamOffsetHours = 7;

  static DateTime getVietnamNow() {
    final now = DateTime.now();
    final utcNow = now.toUtc();
    
    // Convert UTC về Vietnam time (GMT+7)
    return utcNow.add(Duration(hours: vietnamOffsetHours));
  }

  static DateTime toVietnamTime(DateTime utcDateTime) {
    if (utcDateTime.isUtc) {
      return utcDateTime.add(Duration(hours: vietnamOffsetHours));
    }
    // Nếu đã là local time, convert sang UTC trước rồi mới convert về Vietnam
    final utc = utcDateTime.toUtc();
    return utc.add(Duration(hours: vietnamOffsetHours));
  }

  static DateTime toUtcTime(DateTime vietnamDateTime) {
    if (vietnamDateTime.isUtc) {
      return vietnamDateTime.subtract(Duration(hours: vietnamOffsetHours));
    }
    // Nếu là local time, convert sang UTC
    final utc = vietnamDateTime.toUtc();
    return utc.subtract(Duration(hours: vietnamOffsetHours));
  }

  static String toIso8601String(DateTime dateTime) {
    // Đảm bảo DateTime là Vietnam time
    final vietnamTime = dateTime.isUtc 
        ? toVietnamTime(dateTime) 
        : dateTime;
    
    // Format với timezone offset
    final year = vietnamTime.year.toString().padLeft(4, '0');
    final month = vietnamTime.month.toString().padLeft(2, '0');
    final day = vietnamTime.day.toString().padLeft(2, '0');
    final hour = vietnamTime.hour.toString().padLeft(2, '0');
    final minute = vietnamTime.minute.toString().padLeft(2, '0');
    final second = vietnamTime.second.toString().padLeft(2, '0');
    
    return '$year-$month-${day}T$hour:$minute:$second+07:00';
  }

  static DateTime fromBackendString(String dateTimeString) {
    try {
      final parsed = DateTime.parse(dateTimeString);
      
      // Nếu là UTC, convert về Vietnam time
      if (parsed.isUtc) {
        return toVietnamTime(parsed);
      }
      
      // Nếu đã có timezone info, giữ nguyên
      return parsed;
    } catch (e) {
      // Nếu parse lỗi, trả về thời gian hiện tại
      return getVietnamNow();
    }
  }

  static bool isVietnamTime(DateTime dateTime) {
    return !dateTime.isUtc;
  }
}

File: api/Utils/DateTimeHelper.cs
using System;

namespace RealEstateHubAPI.Utils
{
    public static class DateTimeHelper
    {
        private static TimeZoneInfo? _vietnamTimeZone;

        private static TimeZoneInfo GetVietnamTimeZone()
        {
            if (_vietnamTimeZone != null)
            {
                return _vietnamTimeZone;
            }

            try
            {
                // Windows: "SE Asia Standard Time"
                _vietnamTimeZone = TimeZoneInfo.FindSystemTimeZoneById("SE Asia Standard Time");
            }
            catch (TimeZoneNotFoundException)
            {
                try
                {
                    // Linux/macOS: "Asia/Ho_Chi_Minh"
                    _vietnamTimeZone = TimeZoneInfo.FindSystemTimeZoneById("Asia/Ho_Chi_Minh");
                }
                catch (TimeZoneNotFoundException)
                {
                    // Fallback: tạo custom timezone GMT+7
                    _vietnamTimeZone = TimeZoneInfo.CreateCustomTimeZone(
                        "Vietnam Standard Time",
                        TimeSpan.FromHours(7),
                        "Vietnam Standard Time",
                        "Vietnam Standard Time"
                    );
                }
            }

            return _vietnamTimeZone;
        }

        public static DateTime GetVietnamNow()
        {
            var vietnamTimeZone = GetVietnamTimeZone();
            var utcNow = DateTime.UtcNow;
            return TimeZoneInfo.ConvertTimeFromUtc(utcNow, vietnamTimeZone);
        }

        public static DateTime ToVietnamTime(DateTime utcDateTime)
        {
            if (utcDateTime.Kind == DateTimeKind.Unspecified)
            {
                // Nếu không xác định được kind, giả định là UTC
                var utc = DateTime.SpecifyKind(utcDateTime, DateTimeKind.Utc);
                var vietnamTimeZone = GetVietnamTimeZone();
                return TimeZoneInfo.ConvertTimeFromUtc(utc, vietnamTimeZone);
            }

            if (utcDateTime.Kind == DateTimeKind.Utc)
            {
                var vietnamTimeZone = GetVietnamTimeZone();
                return TimeZoneInfo.ConvertTimeFromUtc(utcDateTime, vietnamTimeZone);
            }

            // Nếu đã là local time, trả về nguyên vẹn
            return utcDateTime;
        }

        public static DateTime ToUtcTime(DateTime vietnamDateTime)
        {
            var vietnamTimeZone = GetVietnamTimeZone();
            
            // Nếu DateTime không có Kind, giả định là Vietnam local time
            if (vietnamDateTime.Kind == DateTimeKind.Unspecified)
            {
                var local = DateTime.SpecifyKind(vietnamDateTime, DateTimeKind.Unspecified);
                return TimeZoneInfo.ConvertTimeToUtc(local, vietnamTimeZone);
            }

            if (vietnamDateTime.Kind == DateTimeKind.Local)
            {
                return TimeZoneInfo.ConvertTimeToUtc(vietnamDateTime, vietnamTimeZone);
            }

            // Nếu đã là UTC, trả về nguyên vẹn
            return vietnamDateTime;
        }

        public static bool IsVietnamTime(DateTime dateTime)
        {
            var vietnamTimeZone = GetVietnamTimeZone();
            var offset = vietnamTimeZone.GetUtcOffset(dateTime);
            return offset == TimeSpan.FromHours(7);
        }
    }
}

